Pyodide updates: pyemscripten platform tag (PEP 783), 314.0a1, `pyodide-eol` flag, and maintenance updates (#2812)
pypa/cibuildwheel: docs/_internal/pyodide-maintenance.md

Reviewer: This is nice as a point-in-time capture of the current thinking! I'll just caution that these kinds of maintainer notes often go out-of-date or else require lots of maintenance as time goes on. Two ideas-

- It might be a good idea to chuck a "Last updated: April 2026" note in the file so we know how fresh it is.
- I'd recommend keeping it as short as possible, so we have less to maintain down the road!

But, if you're happy to keep it updated @agriyakhetarpal, and it's helpful for you, this sounds good to me.

# Maintaining Pyodide support

> This page describes some short steps on how to update cibuildwheel's Pyodide platform code when the set of supported Pyodide Python versions changes – for example, when:

Reviewer 2: ```suggestion
This page describes how to update cibuildwheel's Pyodide platform code when either:
```

> - a new Pyodide alpha release arrives with support for a new [Pyodide ABI](https://pyodide.org/en/latest/development/abi.html) (which may be tied to updates in Emscripten and CPython versions, compiler/linker flags, Rust toolchain support and so on), or

Reviewer 3: ```suggestion
- a new Pyodide alpha release arrives with support for a new [PyEmscripten Platform](https://pyodide.org/en/latest/development/abi.html) (which is tied to updates in Emscripten and CPython versions, compiler/linker flags, and so on), or
```

- when that alpha graduates to a stable release.

## Background

> Pyodide ships/may ship two kinds of Python builds at any point in time (note that these numbers may not be up to date):

Reviewer 2: ```suggestion
Pyodide has two types of releases that matter to cibuildheel:
```

- **Stable** – the most recent full Pyodide release (e.g., `0.29.x` / cp313). This is enabled by default with no special `CIBW_ENABLE` flag needed.

> - **Prerelease** – an alpha/beta/rc Pyodide release that uses the _next_ CPython version (e.g., `314.0.0a1` / cp314). Users must opt in with `CIBW_ENABLE: pyodide-prerelease` to build against this version. This may or may not be available at any given time, depending on the Pyodide release cycle.

Reviewer 3: As I mentioned in the previous comment, I think we should remove differentiating Pyodide alpha from stable at least in cibuildwheel-side of view.

The guards in `cibuildwheel/selector.py` enforce this distinction. The constraints files under `cibuildwheel/resources/` pin the exact tool versions that go with each build.

---

## When a new Pyodide prerelease becomes available

For example, consider a scenario when Pyodide ships a new `315.0.0a1` with cp315 support.

### 1. Add the new Python configuration

In `cibuildwheel/resources/build-platforms.toml`, add an entry under `[pyodide]`:

```toml
{ identifier = "cp315-pyodide_wasm32", version = "3.15", default_pyodide_version = "315.0.0a1", node_version = "v24" },
```

> `version` is the CPython version string, `default_pyodide_version` is the Pyodide release to use when the user does not pin one explicitly (use the latest available alpha/beta for a prerelease entry), and `node_version` is the minimum Node.js major required by that Pyodide release — check the [pyodide-build FAQ](https://pyodide-build.readthedocs.io/en/latest/faq.html#what-node-js-version-do-i-need) for a rudimentary idea of what the correct value is.

Reviewer 2: The `node_version` situation is a bit wishy-washy. We should improve our docs on node versions. No action needed here.

Author: Agreed

### 2. Update the prerelease guards in the selector

In `cibuildwheel/selector.py`, update the patterns in the `PyodidePrerelease` guards to match the new identifier:

```python
if EnableGroup.PyodidePrerelease not in self.enable and fnmatch(
    build_id, "cp315-pyodide_*"
):
    return False
```

> Also, update the comment above it to reflect the new version numbers and the instructions for when this guard should eventually be disabled.

Reviewer 2: Probably not needed to update this docs file, I think people can understand the pattern.

### 3. Generate and pin a constraints file

> The easiest way is to run the `update_constraints` `nox` session, which reads `build-platforms.toml` and regenerates all Pyodide constraints files automatically:

Reviewer 2: Let's not say how to run the generator script directly here because the nox code could change and we'd  unlikely to remember to update this. Anyone who wants to know what it does can read the noxfile.
```suggestion
Run the `update_constraints` `nox` session, which reads `build-platforms.toml` and regenerates all Pyodide constraints files automatically:
```

```bash
nox -s update_constraints
```

Alternatively, run the generator script directly:

```bash
python bin/generate_pyodide_constraints.py 315.0.0a1 \
    | uv pip compile - --python-version=3.15 \
        -o cibuildwheel/resources/constraints-pyodide315.txt
```

### 4. Update tests and update CI configuration

- Update the unit tests so the new identifier is accepted by the selector with `PyodidePrerelease` enabled and rejected without it. Pyodide-specific integration tests may also need their hardcoded expected-wheel lists extended.

- Run the full test suite with `CIBW_PLATFORM=pyodide` and `CIBW_ENABLE=pyodide-prerelease` environment variables to make sure the new configuration is well exercised in CI.

> - In GHA, make sure `CIBW_ENABLE: pyodide-prerelease` is set in the Pyodide CI job so that the prerelease identifier is exercised in CI.

Reviewer 2: Does this need to be changed on every update? Surely we should set `CIBW_ENABLE=pyodide-prerelease` all the time, even when there isn't a prerelease?

Author: You're right, no, I was wrong about this – I thought we would raise an error if `CIBW_ENABLE=pyodide-prerelease` is set and no Pyodide pre-release is available to be enabled, in cibuildwheel, but turns out this is a no-op. We silently don't select any pre-release builds if none are available, we're all good. I'm changing it so that it is always set.

## When a Pyodide prerelease becomes stable

Pyodide uses a versioning scheme where the stable release for a given CPython version is named `[PythonMajorMinor].0.0`, so the first stable release shipping cp314 will be **`314.0.0`**. See [pyodide/pyodide#6084](https://github.com/pyodide/pyodide/issues/6084) for a rationale of this versioning scheme.

### 1. Update the stable entries, and remove (or replace) the prerelease entry

> In `build-platforms.toml`, update the former prerelease entry's `default_pyodide_version` to the new stable release (e.g. `314.0.0`) and remove the prerelease marker from the identifier if present. Remove previous prerelease entries if they are now obsolete, or update them to the next prerelease if one is available. Based on your discretion, you may choose to keep the old stable entry for a while if it is still supported by Pyodide, or remove it immediately if it is already retired.

Reviewer 2: Can we simplify this to say just whatever the most likely thing is? We're talking to ourselves here, if we later disagree with the advice we can deviate from it or change it.

Author: I think we need to decide on this before I change this sentence, because we want to be aligned on a policy for how many Pyodide versions/ABIs we should support at a time in cibuildwheel. So far, it's always been two at a time (or even one, some time back). Please read my other comment: https://github.com/pypa/cibuildwheel/pull/2812#discussion_r3073778240

Reviewer: `bin/update_pythons.py` defines a lot of the logic, especially around default_pyodide_version. It might be cleaner when adding a version, to add a stub entry to `build-platforms.toml` and then run the script. If we need some behaviour that's not captured by the script, we'll have to update it, otherwise the `Update Dependencies` PRs will fight it.

### 2. Disable or update the prerelease guards

In `selector.py`:

- **If a new prerelease is available**: update the `fnmatch` pattern to the next identifier (e.g. `cp315-pyodide_*`) as described above.

> - **If there is no new prerelease**L remove the `PyodidePrerelease` logic entirely.

Reviewer 2: We want to comment it right?
```suggestion
- **If there is no new prerelease** comment out the `PyodidePrerelease` logic.
```

### 3. Update the constraints file

Run `nox -s update_constraints` to regenerate the constraints file for the newly stable version. If the entry was already in `build-platforms.toml` as a prerelease, its constraints file already exists, and this step just refreshes it against the stable Pyodide release and updates the dependencies' versions.

### 4. Update tests and CI configuration

Update the unit tests so the newly stable identifier is accepted by the selector without needing `PyodidePrerelease` in the enable set. Pyodide-specific integration tests may also need their hardcoded expected-wheel lists extended.

- Run the full test suite with `CIBW_PLATFORM=pyodide` environment variables to make sure the new configuration is well exercised in CI.

> - Remove `CIBW_ENABLE: pyodide-prerelease` from the GHA Pyodide CI job steps, as no prerelease identifiers should remain after the stable release ships.

Reviewer 2: Why not always set it?

> # Maintaining Pyodide support

Reviewer 3: Thanks for writing this!
